src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1459,6 +1477,7 @@ private async Task HandleConnectionClose(ConnectionState connectionState)
 
             // Cancel any outstanding invocations within the connection lock
             connectionState.CancelOutstandingInvocations(connectionState.CloseException);
+            connectionState.Complete();
```

Comment:
Nulling out `_state.CurrentConnectionStateUnsynchronized` inside the `_connectionLock` semaphore before disposing the `ConnectionContext` was supposed to prevent any race with the transport independently closing right as someone calls `HubConnection.StopAsync()`. I'm guessing this didn't work because we release the semaphore before we start the call to `connectionState.StopAsync()` in `HubConnection.StopAsyncCore`. I think starting the task with the semaphore acquired is probably the proper way to fix the race. It avoids adding the extra `_complete` bool too.

This code could still probably race because `Complete()` is called after disposing the `ConnectionContext` meaning that the call to `Connection.Transport.Input.CancelPendingRead()` in `ConnectionState.StopAsyncCore` could throw an ODE.

Reply:
> I think starting the task with the semaphore acquired is probably the proper way to fix the race. It avoids adding the extra _complete bool too.

Huh... that's a lot simpler. Do you know why we don't just await it inside the semaphore too?